sniffdet.1
.\" libsniffdet - A library for network sniffers detection
.\" Copyright (c) 2002
.\"   Ademar de Souza Reis Jr. <ademar@ademar.org>
.\"   Milton Soares Filho <eu_mil@yahoo.com>
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111,
.\" USA.
.\"
.TH SNIFFDET 1 2002-11-25 "sniffdet manpage" "Remote Sniffer Detection Tool"
.SH NAME
sniffdet - Remote sniffer detection tool
.SH SYNOPSIS
.PP
sniffdet [options] TARGET
.PP
.SH DESCRIPTION
.PP
Sniffdet is an OpenSource implementation of a set of tests for remote
sniffers detection in TCP/IP network environments. It is useful for
remote sniffer detection or to just discover machines which are running
in promiscuous mode.
.PP
Sniffdet is very flexible and allows you to configure many of its options by
using the config file /etc/sniffdet.conf. It also has plugins support for
the result of its tests (currently, XML and stdout output are created).
.PP
You can see the full documentation at http://sniffdet.sourceforge.net
.PP
.SH OPTIONS
TARGET is a canonical hostname or a dotted decimal IPv4 address
.TP
\fB-i  --iface=DEVICE\fR
Use network DEVICE interface for tests.
.br
Default is eth0 in linux systems.
.TP
\fB-l  --log=FILE\fR
Use FILE for tests log.
.br
Default is none
.TP
\fB-c  --configfile=FILE\fR
Use FILE as configuration file for application.
.br
Default is /etc/sniffdet.conf
.TP
\fB-f  --hostsfile=FILE\fR
Use FILE as input for tests target. The file must be in ascii with one
hostname, IP or net address per line. Comments start with '#'
.TP
\fB-u  --uid=UID\fR
Run program with UID (after dropping root).
.br
Default is UID 280 (from config file)
.TP
\fB-g  --gid=GID\fR
Run program with GID (after dropping root)
.br
Default is GID 280 (from config file)
.TP
\fB-t  --test=[testname]\fR
Perform a specific test(s)
.br
Where [testname] is a list composed by at least one of:
    \fBdns\fR         DNS test
    \fBarp\fR         ARP response test
    \fBicmp\fR        ICMP ping response test
    \fBlatency\fR     ICMP ping latency test
.br
.br
See the full documentation included with the
.br
library for information about all tests
.TP
\fB    --pluginsdir=[directory]\fR
Select a directory where sniffdet will load plugins from
.TP
\fB-p  --plugin=[plugin_name]\fR
Select a plugin to load (xml, stdout, etc).
.TP
\fB-f  --targetsfile=[file]\fR
Scan all targets present in a file with a test.
.TP
\fB-v  --verbose\fR
Run in verbose mode (extra output messages).
.br
Default is no.
.TP
\fB-s  --silent\fR
Run in silent mode (no output messages).
.br
Default is no.
.TP
\fB-h, --help\fR
Show a help screen and exit
.TP
\fB--version\fR
Show version information and exit
.PP
.SH EXAMPLES
.PP
# sniffdet -i eth1 -t dns,arp,icmp foo.localdomain
.PP
Test the host foo.localdomain with dns, arp and icmp tests using the
interface eth1
.PP
.PP
# sniffdet -i eth0 -t latency foo.localdomain --plugin=xml
.PP
Test the machine foo.localdomain using the latency test through the
interface eth0. Output results using the xml plugin.
.SH BUGS
This program can be considered in beta stage since it was not widely tested.
Your support is appreciated. :-)
.br
.PP
.PP
Also see our TODO file.
.SH COPYRIGHT
Copyright (c) 2002-2003
   Ademar de Souza Reis Jr. <ademar@ademar.org>
   Milton Soares Filho <eu_mil@yahoo.com>
.SH SEE ALSO
.BR sniffdet.conf (5)
.BR libsniffdet (3)
.PP
.BR http://sniffdet.sourceforge.net


CHANGES
$Id$

sniffdet 0.95
----------------------------------------------------------------------------

This is a bugfix release, intended to fix compilation problems on modern
Linux distributions. The project has been offline for a long time, but
it has not been forgotten. :-)

- new build-system, using state-of-the-art autootols techniques and
  compiling with latest linux distributions :-)
- patch from Andrew Lockhart <andrew [ at ] ingsoc.net>, fixing get_mac()
  which was totally b0rken
- General code cleanup

sniffdet 0.9
----------------------------------------------------------------------------

2003 Jul 04
- Fixed test cancelation using the callback (it was not working in many cases).
- Fixed test status report to the callback.
- Created '/doc' directory. By now, this directory will contain
  only technical documentation (in english).
- Added libsniffdet-usage-example.c to /doc (guess what is it) :)
- Fixed segfault when using a targets file.
- Fixed tmout and tries default values in icmp, dns and arp
  tests.
- Changes in snprintf() calls to avoid format string bugs -
  scan_args(), exploitable only if sniffdet is suid (not standard,
  strongly not recommended).
- sndet_random() is now more robust.
- Fixed issue #645769 (arp test validity could be checked).
  Actually, I'm not just checking the validity, but eliminating
  the false positives possibility that existed before. :)
- Added RCS tags to all .{c,h} files

sniffdet 0.8
----------------------------------------------------------------------------

- Minor code cleanups and interface cosmetic changes
- Fixed bug #645773 (cmdline args and conf_file options mixed)
  The approach I took is a little nasty IMHO (I call scan_args() twice), but
  it works... :-)
- -p/--plugin was not working... fixed.
- Renamed plugins/plugin.h to plugins/plugins.h
- Added a new section to configuration file: plugins{}, with an
  option to set the filename for the xml plugin. In order to do that, I
  changed the plugins API to accept the config structure as one of the
  parameters.
- Fixed bug #662856 (Segmentation fault with CIDR notation
  addresses), reported by Janis Putrams:
  sniffdet was getting sig11 when supplied with a bogus hostname target
  (or something it doesn't understand, like a CIDR network address). The
  problem was in two places: 1) the tests were returning with error without
  setting the info structure. 2) the tests were not returning in all errors,
  specially the latency_test.
- Added sniffdet.conf-debug, with configuration options for debug
  purposes
- Implemented parser for tcp_flags option in latency_test.
- Parallel make doesn't work for sniffdet, so removing it from
  RPM specfile.

sniffdet 0.7 (first public release)
----------------------------------------------------------------------------

2002 Nov 29
- Adding HAVE_SYS_PARAM_H test to configure (it's used to detect OS version,
  among other things).
- Changing latency test default hw address to one which will pass the kernel
  stack as a multicast/broadcast address when in promiscuous mode (0xff,
  0x...)
- Fixing old calls for sleep() (it should be sndet_sleep())
- Moving BUGS file from sndet/src/libs to sndet/ dir and adding some bug
  entries there (and also cleaning TODO file)

2002 Nov 26
- log and callback improvements
- New option: --silent (no output during program execution)
- Added "logtype" entry to configfile (now you can direct log output to one
  or a combination of FILE, SYSLOG, STDERR, STDOUT
- Changed config file entry: "logfile" --> "logfilename"
- using config.h for some variables now (still needs improvements)

2002 Nov 25
- manpage changes (cleanups, cosmetic)
- DEBUG disabled by default (configure)
- Minor cleanups

2002 Nov 22
- Fix for a nasty bug in config_file module... I was using get_int handler
  for a short int variable... BOOM! (no segfault, just variables getting
  lost)
- Some code/documentation cleanup

2002 Oct 31
- Code documentation (comments) in application (mainly config_file.c)
- TODO file improvement
- Minor fixes (typos, etc)

2002 Oct 29
- Config file parser is now fully integraded with application (well...
  almost) :-)

2002 Oct 28
- XML output plugin major rework
- Config file parser major improvements

2002 Oct 25
- Config file parser major improvements
- Manpages filled (sniffdet.1, libsniffdet.3, sniffdet.conf.5)

2002 Oct 07
- setting SYN flag to tcp packet in latency test
- configure should not fail if libdl is not found
  (FreeBSD doesn't have it as a library, but it works)
- threads: s/SA_ONESHOT/SA_RESETHAND/

2002 Oct 02
- config file reading cleanup (still incomplete)
- configure now detects pthread on BSD systems (borrowed from xmms)
- fix app Makefile to recompile sniffdet if the lib changed

2002 Sep 27
- latency test and ping() are now working fine :-)
- minor bugfixes/cleanups

2002 Sep 09
- minor fixes/typos

2002 Aug 19
- config file reading work started

2002 Aug 09
- output plugin preliminary support
- minor bugfixes/enhancements

2002 Aug 02
- test various hosts (loop, output, target list file, etc)
- Build minor improvements (tags handling by Makefiles)

2002 Aug 01
- little cleanup, read hostnames file partially implemented

2002 Jul 15
- little build improvement (debug extra CFLAGS)

2002 Jul 04
- sf.net bug #574668 (dns packet is being trusted)
  The function now checks the pkt_len and returns false (0)
  if the packet is bigger than its real size
- sf.net bug #574676 (PCAP_FILTER_BUFF_SIZE redefinitions)
  PCAP_FILTER_BUFF_SIZE is ok, there are other variables though.
- Compilation fixes

2002 Jun 27
- make install and specfile are now functional

2002 Jun 24

src/sniffdet.c
/*
 *  sniffdet - A tool for network sniffers detection
 *  Copyright (c) 2002
 *      Ademar de Souza Reis Jr. <ademar@ademar.org>
 *      Milton Soares Filho <eu_mil@yahoo.com>
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; version 2 dated
 *  June, 1991.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *  See file COPYING for details.
 *
 *  $Id$
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <limits.h>
#include <getopt.h>
#include <dlfcn.h>
#include <string.h>

#include "config.h"
#include "libsniffdet.h"
#include "sniffdet.h"
#include "log.h"
#include "util.h"

// function prototypes
static int tests_msg_callback(struct test_status *status,
		const int msg_type, char *msg);
static void set_global_defaults(void);
static void scan_args(int argc, char **argv);
static int parse_testnames(char *names);
void sighandler_sigint(void);
static void show_version(void);
static void show_usage(void);
static void show_help(void);

/* static data */
// XXX: use a global structure to hold that?
// XXX: I don't think it's a good idea
static struct arguments args;
static struct snd_tests run_tests;
static int logfd;
struct config_options config;

// global switch used when we want to cancel a test
static int cancel_tests = 0;


/* This is our callback. It receives status notifications from the tests. It
 * can be used to avoid the freezing sensation that we get when running a
 * test for a long time.
 *
 * TODO:
 * - Implement it in a clean, simple and eyecandy way... :-)
 */

static int tests_msg_callback(struct test_status *status,
		const int msg_type, char *msg)
{
#ifdef DEBUG_CALLBACK
	static char *msgs[7] = {
		"RUNNING",
		"NOTIFICATION",
		"ERROR",
		"WARNING",
		"DETECTION",
		"ENDING",
		NULL
	};

	printf("-CB- MSG TYPE: %s\n", msgs[msg_type]);
	if (msg != NULL)
		printf("-CB- Message: %s\n", msg);
	printf("-CB- Percentage: %d%%\n", status->percent);
	printf("-CB- Bytes Sent: %d\n", status->bytes_sent);
	printf("-CB- Bytes Recvd: %d\n", status->bytes_recvd);
#endif

	// avoid warning
	(void) status;

	// just return if we're not using a log
	if (config.global.logtype == LOG_NOLOG)
		return cancel_tests ? 1 : 0;

	// do we have a message?
	if (msg == NULL)
		return cancel_tests ? 1 : 0;

	switch (msg_type) {
		case RUNNING:
			if (config.global.verbose)
				mylog(config.global.logtype, logfd, "%s", msg);
			break;
		case NOTIFICATION:
		case ENDING:
			mylog(config.global.logtype, logfd, "%s", msg);
			break;
		case ERROR:
			mylog(config.global.logtype, logfd, "Error: %s", msg);
			break;
		case WARNING:
			mylog(config.global.logtype, logfd, "Warning: %s", msg);
			break;
	}

	return cancel_tests ? 1 : 0;
}

/* TODO:
 * Our main function is big... We could split it in several little
 * static functions and call them (like the handling of config file,
 * plugins, etc).
 */
int main(int argc, char **argv)
{
	struct test_info t_info[5]; // 4 tests + ending
	struct sndet_device *device;
	char errbuf[LIBSNIFFDET_ERR_BUF_LEN];
	char plugin_path[PATH_MAX];
	int test = 0;
	int i;
	int status;
	char **targets = NULL;
	FILE *f_hosts;

	// output plugin
	void *o_plugin;
	int (*test_output)(char *, struct test_info *, struct config_options, char *);

	// set everything in config struct to NULL (zero)
	memset(&config, 0, sizeof(struct config_options));

	// set default options and general info
	set_global_defaults();

	// scan command line arguments
	scan_args(argc, argv);

	// read/parse config file
	if (read_config(args.configfile)) {
		fprintf(stderr, "Error reading config file %s\n",
				args.configfile);
		fprintf(stderr, "Exiting...\n");
		return 1;
	}

	// scan command line arguments (again to override config file options)
	// TODO: use a better approach
	scan_args(argc, argv);

	// open log file
	if ((config.global.logfilename[0] != '\0') &&
			(config.global.logtype & LOG_USE_FILE)) {
		logfd = open(config.global.logfilename, O_APPEND|O_WRONLY|O_CREAT, 0644);
		if (logfd == -1) {
			fprintf(stderr, "Error opening log file: %s\n",
					config.global.logfilename);
			perror("");
			fprintf(stderr, "Log disabled\n");
			config.global.logfilename[0] = '\0';
		}
	}

	/* TODO:
	 * check geteuid() and warn the user if the file is suid,
	 * since I think it's not a good idea to have sniffdet suid in the
	 * system
	 */
	// running as root?
	//
	if (getuid() != 0) {
		mylog(config.global.logtype | LOG_USE_STDERR, logfd,
				"This program needs root privileges to run!");
		mylog(config.global.logtype | LOG_USE_STDERR, logfd,
				"Exiting...");
		return 1;
	}

	// initialize device(s)
	// (must be root to do that)
	if ((device = sndet_init_device(config.global.iface, 1, errbuf)) == NULL) {
		mylog(config.global.logtype | LOG_USE_STDERR, logfd,
				"Error initializing interface %s", config.global.iface);
		mylog(config.global.logtype | LOG_USE_STDERR, logfd,
				"%s", errbuf);
		mylog(config.global.logtype | LOG_USE_STDERR, logfd,
				"Exiting...");
		return 1;
	}

	mylog(config.global.logtype, logfd,
			"%s", "--- sniffdet session started ---");

	// bump some network info to the log
	{
	struct in_addr net[2];
	char temp_buf[256];

	net[0].s_addr = device->network;
	net[1].s_addr = device->netmask;
	// inet_ntoa() returns a statically buffer, which
	// is overwrited at every call - see manpage
	snprintf(temp_buf, 256, "%s", inet_ntoa(net[1]));

	mylog(config.global.logtype, logfd,
			"NETWORK INFO: device: %s; network: %s (0x%X); netmask: %s (0x%X)",
			device->device, inet_ntoa(net[0]), net[0].s_addr,
			temp_buf, net[1].s_addr);
	}

	// read targets file
	if (args.targetsfile != NULL) {
		f_hosts = fopen(args.targetsfile, "r");
		if (f_hosts == NULL) {
			mylog(config.global.logtype | LOG_USE_STDERR, logfd,
					"Error opening targets file: %s", args.targetsfile);
			mylog(config.global.logtype | LOG_USE_STDERR, logfd,
					strerror(errno));
			mylog(config.global.logtype | LOG_USE_STDERR, logfd,
					"Exiting...");
			return 1;
		}
		targets = parse_targets_file(f_hosts);
		if (targets == NULL)
			DABORT();

		fclose(f_hosts);
	}
	else {
		targets = malloc(sizeof (char *) * 2);
		targets[0] = args.target;
		targets[1] = NULL;
	}

	// now we run as some unprivileged user
	if (config.global.UID == 0) {
		mylog(config.global.logtype | LOG_USE_STDERR, logfd,
				"WARNING: Running as root user!");
		mylog(config.global.logtype | LOG_USE_STDERR, logfd,
				"         Check your configuration!");
	}
	else {
		if (config.global.verbose) {
			mylog(config.global.logtype, logfd,
					"Dropping root privileges UID: %d, GID: %d",
					config.global.UID, config.global.GID);
		}
		if (!drop_root(config.global.UID, config.global.GID)) {
			mylog(config.global.logtype | LOG_USE_STDERR, logfd,
					"Error dropping privileges");
			mylog(config.global.logtype | LOG_USE_STDERR, logfd,
					"Exiting...");
			return 1;
		}
	}

	// open output plugin
	snprintf(plugin_path, PATH_MAX, "%s/%s", config.global.plugins_dir, config.global.plugin);
	if (config.global.verbose) {
		mylog(config.global.logtype, logfd,
			"Opening plugin: %s\n", plugin_path);
	}
	o_plugin = dlopen(plugin_path, RTLD_LAZY);
	if (!o_plugin) {
			mylog(config.global.logtype | LOG_USE_STDERR, logfd,
					"Error loading plugin: %s", plugin_path);
			mylog(config.global.logtype | LOG_USE_STDERR, logfd,
					"%s", dlerror());
			mylog(config.global.logtype | LOG_USE_STDERR, logfd,
					"Exiting...");
			return 1;
	}
	test_output = dlsym(o_plugin, "test_output");
	if (!o_plugin) {
			mylog(config.global.logtype | LOG_USE_STDERR, logfd,
					"Invalid plugin - %s", plugin_path);
			mylog(config.global.logtype | LOG_USE_STDERR, logfd,
					"%s", dlerror());
			mylog(config.global.logtype | LOG_USE_STDERR, logfd,
					"Exiting...");
			return 1;
	}

	/* loop to test every target */
	for (i = 0; targets[i] != NULL; i++) {

		/* call tests */
		test = 0;

		if (!sndet_resolve(targets[i])) {
			mylog(config.global.logtype | LOG_USE_STDERR, logfd,
					"Warning: Cannot resolve hostname \"%s\" [ignoring host]",
					targets[i]);
			continue;
		}

		// ICMP test
		if (run_tests.icmptest && !cancel_tests) {
			mylog(config.global.logtype, logfd,
					"Calling ICMP Test for %s on %s",
					targets[i], device->device);
			status = sndet_icmptest(targets[i],
					device,
					config.icmptest.timeout, // timeout (secs)
					config.icmptest.tries, // tries
					config.icmptest.interval, // interval (msecs)
					tests_msg_callback,
					&t_info[test++],
					config.icmptest.fake_hwaddr); // fake mac address
			mylog(config.global.logtype, logfd,
					"%s: %s (%d)", "ICMP Test done",
					(status ? "failure" : "sucess"), status);
		}

		// DNS
		if (run_tests.dnstest && !cancel_tests) {
			mylog(config.global.logtype, logfd,
					"Calling DNS Test for %s on %s",
					targets[i], device->device);
			status = sndet_dnstest(targets[i],
					device,
					config.dnstest.timeout, // timeout (secs)
					config.dnstest.tries, // tries
					config.dnstest.interval, // interval (msecs)
					tests_msg_callback,
					&t_info[test++],
					// optional data (packet data)
					config.dnstest.fake_ipaddr, // fake ip
					config.dnstest.fake_hwaddr, // fake mac
					config.dnstest.dport, // destination port
					config.dnstest.sport, // source port
					config.dnstest.payload, // payload
					config.dnstest.payload_len); // payload len
			mylog(config.global.logtype, logfd,
					"%s: %s (%d)", "DNS Test done",
					(status ? "failure" : "sucess"), status);
		}

		// ARP
		if (run_tests.arptest && !cancel_tests) {
			mylog(config.global.logtype, logfd,
					"Calling ARP Test for %s on %s",
					targets[i], device->device);
			status = sndet_arptest(targets[i],
					device,
					config.arptest.timeout, // timeout (secs)
					config.arptest.tries, // tries
					config.arptest.interval, // interval (msecs)
					tests_msg_callback,
					&t_info[test++],
					config.arptest.fake_hwaddr); // fake mac address
			mylog(config.global.logtype, logfd,
					"%s: %s (%d)", "ARP Test done",
					(status ? "failure" : "sucess"), status);
		}

		// LATENCY
		if (run_tests.latencytest && !cancel_tests) {

			mylog(config.global.logtype, logfd,
					"Calling Latency Test for %s on %s",
					targets[i], device->device);

			// TODO/FIXME:
			// build bogus_pkt with defaults
			// (and use as much options as possible from the config file)
			status = sndet_latencytest_pktflood(
					targets[i],
					device,
					config.latencytest.timeout, // timeout (sec)
					// interval between measures (msec)
					config.latencytest.probe_interval,
					tests_msg_callback,
					&t_info[test++],
					NULL); // bogus_pkt

			mylog(config.global.logtype, logfd,
					"%s: %s (%d)", "Latency Test done",
					(status ? "failure" : "sucess"), status);
		}

		// mark the end of tests info structure vector
		t_info[test].code = MAX_TESTS;

		// output tests info/result
		// this comes from a plugin
		if ((*test_output)(targets[i], t_info, config, errbuf))
			mylog(config.global.logtype | LOG_USE_STDERR, logfd,
					"Error: %s", errbuf);

	} // end targets loop

	// close plugin
	dlclose(o_plugin);

	// finish/close our device(s)
	mylog(config.global.logtype, logfd,
			"Closing device %s", device->device);

	if (sndet_finish_device(device, errbuf))
		mylog(config.global.logtype | LOG_USE_STDERR, logfd,
				"Error: %s", errbuf);

	mylog(config.global.logtype, logfd,
			"%s", "--- sniffdet session ended ---\n");

	// close log file
	if (config.global.logfilename[0] != '\0')
		close(logfd);

	// free targets string list
	if (targets != NULL) {
		free_stringlist(targets);
	}

	return 0;
}

/* fill some defaults in the args and global structures */
static void set_global_defaults(void)
{
	snprintf(config.global.iface, MAX_CFG_VAR_SIZE, "%s", "eth0");
	config.global.UID = SNDET_DEFAULT_UID; // just a non meaning number
	config.global.GID = SNDET_DEFAULT_GID; // just a non meaning number
	config.global.verbose = 0;
	config.global.silent = 0;
	config.global.logtype = LOG_NOLOG;
	config.global.logfilename[0] = '\0';
	snprintf(config.global.plugins_dir, MAX_CFG_VAR_SIZE, "%s", SNDET_PLUGINSDIR);
	snprintf(config.global.plugin, MAX_CFG_VAR_SIZE, "%s", "stdout.so");

	snprintf(config.plugins.xml.filename, MAX_CFG_VAR_SIZE, "%s", "tests_result.xml");

	args.target = NULL;
	args.targetsfile = NULL;
	args.configfile = SNDET_CONFIG;

	// no tests by default
	run_tests.dnstest = 0;
	run_tests.icmptest = 0;
	run_tests.arptest = 0;
	run_tests.latencytest = 0;

	// no log by default
	logfd = -1;

	return;
}


/* scan_args()
 * call getopt*() to process the command line arguments
 * passed. Fills a global struct with some info.
 *
 * Notice we call this function a second time after we read the config file
 * (yes, I agree, this is nasty, but it is used to override config options
 * with command line arguments...)
 * TODO: Use a better approach :-)
 */
static void scan_args(int argc, char **argv)
{
	static int pass = 0; // 1 if this function was already called
	int i;
	int mandatory_testname_option = 0;
	int option_index;
	static struct option long_options[] = {
		{"help", 0, NULL, 'h'},
		{"version", 0, NULL, 1000},
		{"verbose", 0, NULL, 'v'},
		{"silent", 0, NULL, 's'},
		{"log", 1, NULL, 'l'},
		{"configfile", 1, NULL, 'c'},
		{"targetsfile", 1, NULL, 'f'},
		{"pluginsdir", 1, NULL, 1001},
		{"plugin", 1, NULL, 'p'},
		{"iface", 1, NULL, 'i'},
		{"test", 1, NULL, 't'},
		{"uid", 1, NULL, 'u'},
		{"gid", 1, NULL, 'g'},
		{0, 0, 0, 0}
	};

	// since we call this function more than once, we
	// have to reset the global variables it has
	if (pass)
		optind = opterr = optopt = 0;

	// set program name
	args.prgname = argv[0];

	while ((i = getopt_long(argc, argv, "hsvl:p:c:f:i:t:u:g:", long_options,
					&option_index)) != EOF) {
		switch (i) {
			case 'h':
				show_help();
				exit(0);
				break;
			case 'v':
				config.global.verbose = 1;
				break;
			case 's':
				config.global.silent = 1;
				break;
			case 'l':
				config.global.logtype |= LOG_USE_FILE;
				snprintf(config.global.logfilename, MAX_CFG_VAR_SIZE, "%s", optarg);
				break;
			case 1001:
				snprintf(config.global.plugins_dir, MAX_CFG_VAR_SIZE, "%s", optarg);
				break;
			case 'p':
				snprintf(config.global.plugin, MAX_CFG_VAR_SIZE, "%s", optarg);
				break;
			case 'i':
				snprintf(config.global.iface, MAX_CFG_VAR_SIZE, "%s", optarg);
				break;
			case 'u':
				config.global.UID = atoi(optarg);
				break;
			case 'g':
				config.global.GID = atoi(optarg);
				break;
			case 'c':
				if (pass)
					break;
				args.configfile = strdup(optarg);
				break;
			case 'f':
				if (pass)
					break;
				args.targetsfile = strdup(optarg);
				break;
			case 't':
				if (pass)
					break;
				mandatory_testname_option = 1;
				if (parse_testnames(optarg) == 0) {
					fprintf(stderr, "Error: Invalid test required!\n");
					show_usage();
					exit(-1);
				}
				break;
			case 1000:
				show_version();
				exit(0);
				break;
			case '?': /* invalid argument */
				show_usage();
				exit(-1);
		}
	}

	if (pass)
		return;

	if (!mandatory_testname_option) {
		fprintf(stderr, "Error: No tests to perform!\n");
		show_usage();
		exit(-1);
	}

	/* get target */
	if (optind < argc) {
		args.target = strdup(argv[optind]);
	}
	else {
		if (args.targetsfile == NULL) { // not using a targetsfile destination
			fprintf(stderr, "Error: No destination target!\n");
			show_usage();
			exit(-1);
		}
	}

	// set pass to 1 (to inform we already called this function)
	pass = 1;
}


/* look for tests to perform in a string
 * and set the run_tests flag
 *
 * TODO:
 * Use an approach like the one in the config_file module, where we have a
 * structure with all strings and handlers for them. It would be easier to
 * add new tests using that approach.
 */
static int parse_testnames(char *names)
{
	int count = 0;

	if (strstr(names, "dns")) {
		run_tests.dnstest = 1;
		count++;
	}
	if (strstr(names, "icmp")) {
		run_tests.icmptest = 1;
		count++;
	}
	if (strstr(names, "arp")) {
		run_tests.arptest = 1;
		count++;
	}
	if (strstr(names, "latency")) {
		run_tests.latencytest = 1;
		count++;
	}

	return count;
}

/* Copyrigth notice and release
 */
static void show_version(void)
{
	printf("sniffdet %s\n", PACKAGE_VERSION);
	printf("A Remote Sniffer Detection Tool\n");
	printf("Copyright (c) 2002\n");
	printf("   Ademar de Souza Reis Jr. <ademar@ademar.org>\n");
	printf("   Milton Soares Filho <eu_mil@yahoo.com>\n");
}


/* Helps the user about the syntax of calling this program
 */
static void show_usage(void)
{
	fprintf(stderr, "Usage: %s [options] TARGET\n", args.prgname);
	fprintf(stderr, "try '%s --help' for more information\n", args.prgname);
}


/* Helps the user about how to call this program
 */
static void show_help(void)
{
	show_version();
	printf("\n");
	printf("Usage: %s [options] TARGET\n", args.prgname);
	printf("  Where:\n");
	printf("  TARGET is a canonical hostname or a dotted decimal IPv4 address\n");
	printf("\n");
	printf("  -i  --iface=DEVICE     Use network DEVICE interface for tests\n");
	printf("  -c  --configfile=FILE  Use FILE as configuration file\n");
	printf("  -l  --log=FILE         Use FILE for tests log\n");
	printf("  -f  --targetsfile=FILE Use FILE for tests target\n");
	printf("      --pluginsdir=DIR   Search for plugins in DIR\n");
	printf("  -p  --plugin=FILE      Use FILE plugin\n");
	printf("  -u  --uid=UID          Run program with UID (after dropping root)\n");
	printf("  -g  --gid=GID          Run program with GID (after dropping root)\n");
	printf("\n");
	printf("  -t  --test=[testname]  Perform specific test\n");
	printf("      Where [testname] is a comma-separated list of the following:\n");
	printf("        dns         DNS test\n");
	printf("        arp         ARP response test\n");
	printf("        icmp        ICMP ping response test\n");
	printf("        latency     ICMP ping latency test\n");
	printf("\n");
	printf("  -s  --silent           Run in silent mode (no output, only call plugin with results)\n");
	printf("  -v  --verbose          Run in verbose mode (extended output)\n");
	printf("  -h, --help             Show this help screen and exit\n");
	printf("      --version          Show version info and exit\n");
	printf("\n");
	printf("Defaults:\n");
	printf("    Interface: \"eth0\"\n");
	printf("    Log file: \"sniffdet.log\"\n");
	printf("    Config file: \"%s\"\n", SNDET_CONFIG);
	printf("    Plugins Directory: \"%s\"\n", SNDET_PLUGINSDIR);
	printf("    Plugin: \"stdout.so\"\n");
	printf("\n");
	printf("You have to specify at least one test to be used (there's no default)\n");
}
